extensions/amp-selector/0.2/selector.js

```diff
@@ -0,0 +1,127 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as CSS from './selector.css';
+import * as Preact from '../../../src/preact';
+import {useState} from '../../../src/preact';
+
+const SelectorContext = Preact.createContext({});
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Selector(props) {
+  const {
+    'value': value,
+    'defaultValue': defaultValue,
+    'disabled': disabled,
+  } = props;
+  const isMultiple = props['multiple'] !== undefined;
+  const [selectedState, setSelectedState] = useState(
+    value ? [].concat(value) : defaultValue ? [].concat(defaultValue) : []
+  );
+  // TBD: controlled values require override of properties.
+  const selected = /** @type {!Array} */ (value
+    ? [].concat(value)
+    : selectedState);
+  const selectOption = (option) => {
+    if (!option) {
+      return;
+    }
+    const {'onChange': onChange} = props;
+    let newValue = null;
+    if (isMultiple) {
+      newValue = selected.includes(option)
+        ? selected.filter((v) => v != option)
+        : selected.concat(option);
+    } else if (!selected.includes(option)) {
+      newValue = [option];
+    }
+    if (newValue) {
+      setSelectedState(newValue);
+      if (onChange) {
+        onChange({
+          target: {value: isMultiple ? newValue : newValue[0], option},
+        });
+      }
+    }
+  };
+
+  const tag = props['tagName'] || 'div';
+  // TODO: Support '.' access to return <props.tagName ...> in JSX
+  return Preact.createElement(
+    tag,
+    {
+      ...props,
+      'aria-disabled': disabled,
+      'aria-multiselectable': isMultiple,
+      role: props['role'] || 'listbox',
+    },
+    <SelectorContext.Provider
+      value={{
+        selected,
+        selectOption,
+      }}
+    >
+      {props['children']}
+    </SelectorContext.Provider>
+  );
+}
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Option(props) {
+  const {'option': option, 'disabled': disabled, 'style': style} = props;
+  const getOption = props['getOption'] || (() => option);
+  const selectorContext = Preact.useContext(SelectorContext);
+  const {'selected': selected, 'selectOption': selectOption} = selectorContext;
+  const isSelected = /** @type {!Array} */ (selected).includes(option);
+  const statusStyle = disabled
+    ? CSS.DISABLED
+    : isSelected
+    ? CSS.SELECTED
+    : CSS.OPTION;
+  const optionProps = {
+    ...props,
```

Comment:
can we do `const {role, etc, ...rest} = props` and use `...rest` here? This would be cleaner as it will signify which props are "special" vs "passthrough".

Reply:
Getting the following from eslint: `Collecting props using a rest property is not allowed.eslint(local/no-rest)`

Reply:
@caroqliu When we need to propagate as in `<div {...rest}>`, for now let's do manual `const rest = omit(props, [...props to remove...])`. Where "props to remove" would normally be whatever props we read using `const {...} = props`.

@jridgewell I think the pattern of straight-through propagation is important. We can override, but can we make an exception for this?